speculative_decoding/packages/spec_dec.py

```diff
@@ -0,0 +1,369 @@
+"""Helpers and control flow for draft-target-model speculatice decoding.
+
+Notes:
+* Draft and target model can use different tokenizers, clear text is used as exchange
+  between them. In fact this implementaiton does not optimize for same encoding, it
+  always goes via clear text encoding/decoding (which could be skipped and simplified
+  for same tokenizers). There are checks in place for the situation that a string gets
+  differently tokenized when appending more charaters.
+* There are some limitations of non-zero temperature sampling that could lead to
+  deviating results from direct generation from the target model. This is due to using
+  "hard" draft tokens and not draft probabilities and might be improved in the future.
+
+
+TODO:
+* Change from returning context+generated text to only returning generated.
+* Take another look at stop/break criteria for the generation loop.
+* Maybe clean up debug/timing leftovers.
+* Add caching for word lists.
+"""
+
+import asyncio
+import functools
+import os
+import time
+from typing import Sequence
+
+import colorama
+import helpers  # From packages.
+import numpy as np
+import transformers
+import tritonclient.grpc.aio as triton_grpc
+
+
+class SpeculationState:
```

Comment:
Doc comment will be useful here, like what do we keep track of here, whether it is per request etc.